File: SPRING/Modern_Hospital_Management_System/src/main/java/emranhss/com/Modern_Hospital_Management_System/entity/PharmacySale.java
package emranhss.com.Modern_Hospital_Management_System.entity;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Data
@Entity
@Table(name = "pharmacy_sales")
@NoArgsConstructor
@AllArgsConstructor
public class PharmacySale {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, unique = true)
    private String saleInvoiceNo;

    private String patientType;

    private String patientName;

    private String patientPhone;

    private Long patientId;

    private Long doctorId;

    private String doctorName;

    private Long prescriptionId;

    private Double totalAmount = 0.0;

    private Double discount = 0.0;

    private Double vat = 0.0;

    private Double netPayable = 0.0;

    private Double paidAmount = 0.0;

    private Double changeAmount = 0.0;

    private String paymentMethod = "CASH";

    @Column(nullable = false)
    private String paymentStatus = "PAID";

    private String saleType = "COUNTER";

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "billing_id")
    @JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
    private Billing billing;

    @OneToMany(mappedBy = "pharmacySale", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonIgnoreProperties("pharmacySale")
    private List<PharmacySaleItem> items = new ArrayList<>();

    @CreationTimestamp
    @Column(updatable = false)
    private LocalDateTime saleDate;
}

File: Flutter_Hospital_Management/lib/models/pharmacy_sale.dart
class PharmacySaleItem {
  final int? id;
  final int? medicineStockId;
  final String? medicineName;
  final String? batchNumber;
  final int? quantity;
  final double? unitPrice;
  final double? discount;
  final double? subtotal;

  const PharmacySaleItem({
    this.id,
    this.medicineStockId,
    this.medicineName,
    this.batchNumber,
    this.quantity,
    this.unitPrice,
    this.discount,
    this.subtotal,
  });

  factory PharmacySaleItem.fromJson(Map<String, dynamic> j) => PharmacySaleItem(
        id: j['id'],
        medicineStockId: j['medicineStockId'],
        medicineName: j['medicineName'],
        batchNumber: j['batchNumber'],
        quantity: j['quantity'],
        unitPrice: (j['unitPrice'] as num?)?.toDouble(),
        discount: (j['discount'] as num?)?.toDouble(),
        subtotal: (j['subtotal'] as num?)?.toDouble(),
      );
}

class PharmacySale {
  final int id;
  final String? saleInvoiceNo;
  final String? patientType;
  final String? patientName;
  final String? patientPhone;
  final int? patientId;
  final String? doctorName;
  final double? totalAmount;
  final double? discount;
  final double? vat;
  final double? netPayable;
  final double? paidAmount;
  final String? paymentMethod;
  final String? paymentStatus;
  final String? saleType;
  final List<PharmacySaleItem> items;
  final String? saleDate;

  const PharmacySale({
    required this.id,
    this.saleInvoiceNo,
    this.patientType,
    this.patientName,
    this.patientPhone,
    this.patientId,
    this.doctorName,
    this.totalAmount,
    this.discount,
    this.vat,
    this.netPayable,
    this.paidAmount,
    this.paymentMethod,
    this.paymentStatus,
    this.saleType,
    this.items = const [],
    this.saleDate,
  });

  factory PharmacySale.fromJson(Map<String, dynamic> j) => PharmacySale(
        id: j['id'],
        saleInvoiceNo: j['saleInvoiceNo'],
        patientType: j['patientType'],
        patientName: j['patientName'],
        patientPhone: j['patientPhone'],
        patientId: j['patientId'],
        doctorName: j['doctorName'],
        totalAmount: (j['totalAmount'] as num?)?.toDouble(),
        discount: (j['discount'] as num?)?.toDouble(),
        vat: (j['vat'] as num?)?.toDouble(),
        netPayable: (j['netPayable'] as num?)?.toDouble(),
        paidAmount: (j['paidAmount'] as num?)?.toDouble(),
        paymentMethod: j['paymentMethod'],
        paymentStatus: j['paymentStatus'],
        saleType: j['saleType'],
        items: (j['items'] as List?)
                ?.map((e) =>
                    PharmacySaleItem.fromJson(e as Map<String, dynamic>))
                .toList() ??
            [],
        saleDate: j['saleDate']?.toString(),
      );
}
